src/components/DecisioningEngine/createContextProvider.js

```diff
@@ -9,16 +9,77 @@ the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTA
 OF ANY KIND, either express or implied. See the License for the specific language
 governing permissions and limitations under the License.
 */
-export default ({ eventRegistry }) => {
-  const globalContext = {}; // holder of global context like current time/date, browser name/version, day of week, scroll position, time on page, etc
+import getBrowser from "../../utils/getBrowser";
+import parseUrl from "../../utils/parseUrl";
+
+export default ({ eventRegistry }, currentWindow) => {
+  const initialTimeStamp = new Date();
+  const getPageLoadTime = () => {
+    const pageLoadTime = initialTimeStamp.getTime();
+    return {
+      pageLoadTime
+    };
+  };
+  const getBrowserContext = () => {
+    return {
+      name: getBrowser(currentWindow)
+    };
+  };
+  const getPageContext = () => {
+    return {
+      title: currentWindow.title,
+      url: currentWindow.url,
+      ...parseUrl(currentWindow.url)
+    };
+  };
+
+  const getReferrerContext = () => {
+    return {
+      url: currentWindow.referrer,
+      ...parseUrl(currentWindow.referrer)
+    };
+  };
+  const getTimeContext = () => {
+    const newTimeStamp = new Date();
+    return {
+      currentTimestamp: newTimeStamp.getTime(),
+      currentDate: newTimeStamp.getDate(),
+      currentDay: newTimeStamp.getDay(),
+      currentHour: newTimeStamp.getHours(),
+      currentMinute: newTimeStamp.getMinutes(),
+      currentMonth: newTimeStamp.getMonth(),
+      currentYear: newTimeStamp.getFullYear(),
+      pageVisitDuration: newTimeStamp.getTime() - getPageLoadTime().pageLoadTime
+    };
+  };
+
+  const getWindowContext = () => {
+    const height = currentWindow.height;
+    const width = currentWindow.width;
+    const scrollY = currentWindow.scrollY;
+    const scrollX = currentWindow.scrollX;
+    return {
+      height,
+      width,
+      scrollY,
+      scrollX
+    };
+  };
+
   const getContext = addedContext => {
```

Comment:
The way you have it written, `getContext` re-calculates all values each time it is called -- including ones we don't need to re-compute such as browser, page, and referringPage.  I recommend changing it so that it only computes what is necessary when getContext is called.  For example, here is an implementation that computes browser, page and referringPage context and stores it in `coreGlobalContext` -- only at time of initialization of the context provider (once).  Then it reuses the `coreGlobalContext` each time getContext is called, and adds only the values it must calculate to the context.  What do you think?  👇 


```js
export default ({ eventRegistry, window }) => {
  const pageLoadTimestamp = new Date().getTime();

  const getBrowserContext = () => {
    return {
      name: getBrowser(window)
    };
  };
  const getPageContext = () => {
    return {
      title: window.title,
      url: window.url,
      ...parseUrl(window.url)
    };
  };

  const getReferrerContext = () => {
    return {
      url: window.referrer,
      ...parseUrl(window.referrer)
    };
  };
  const getTimeContext = () => {
    const now = new Date();
    const currentTimestamp = now.getTime();

    return {
      pageLoadTimestamp,
      currentTimestamp,
      currentDate: now.getDate(),
      currentDay: now.getDay(),
      currentHour: now.getHours(),
      currentMinute: now.getMinutes(),
      currentMonth: now.getMonth(),
      currentYear: now.getFullYear(),
      pageVisitDuration: currentTimestamp - pageLoadTimestamp
    };
  };

  const getWindowContext = () => {
    const height = window.height;
    const width = window.width;
    const scrollY = window.scrollY;
    const scrollX = window.scrollX;
    return {
      height,
      width,
      scrollY,
      scrollX
    };
  };

  const coreGlobalContext = {
    browser: getBrowserContext(),
    page: getPageContext(),
    referringPage: getReferrerContext()
  };

  const getGlobalContext = () => {
    return {
      ...coreGlobalContext,
      ...getTimeContext(),
      window: getWindowContext()
    };
  };

  const getContext = addedContext => {
    return {
      ...getGlobalContext(),
      ...addedContext,
      events: eventRegistry.toJSON()
    };
  };
  return {
    getContext
  };
};
```

Reply:
applied the fix